src/io_callbacks.zig

```diff
@@ -0,0 +1,463 @@
+                error.SignalInterrupt => continue,
+                else => return err,
+            };
+            self.submitted -= completed;
+            for (cqes[0..completed]) |cqe| {
+                const completion = @intToPtr(*Completion, @intCast(usize, cqe.user_data));
+                completion.result = cqe.res;
+                completion.next = null;
+                // We do not run the completion here (instead appending to a linked list):
+                // * to avoid recursion through `flush_submissions()` and `flush_completions()`,
+                // * to avoid unbounded stack usage, and
+                // * to avoid confusing stack traces.
+                self.completed.push(completion);
+            }
+            if (completed < cqes.len) break;
+        }
+    }
+
+    fn flush_submissions(self: *IO, wait: bool) !void {
+        var wait_nr: u32 = if (wait) 1 else 0;
+        while (true) {
+            wait_nr = std.math.min(self.queued + self.submitted, wait_nr);
+            _ = self.ring.submit_and_wait(wait_nr) catch |err| switch (err) {
+                error.SignalInterrupt => continue,
+                // Wait for some completions and then try again:
+                // See https://github.com/axboe/liburing/issues/281 re: error.SystemResources.
+                // Be careful also that copy_cqes() will flush before entering to wait (it does):
+                // https://github.com/axboe/liburing/commit/35c199c48dfd54ad46b96e386882e7ac341314c5
+                error.CompletionQueueOvercommitted, error.SystemResources => {
+                    try self.flush_completions(true);
+                    continue;
+                },
+                else => return err,
+            };
+            self.submitted += self.queued;
+            self.queued = 0;
+            break;
+        }
+    }
+
+    fn get_sqe(self: *IO, completion: *Completion) void {
+        const sqe = self.ring.get_sqe() catch |err| switch (err) {
+            error.SubmissionQueueFull => {
+                completion.next = null;
+                self.unqueued.push(completion);
+                return;
+            },
+        };
+        self.queued += 1;
+        completion.complete_prep(sqe);
+    }
+
+    /// This struct holds the data needed for a single io_uring operation
+    pub const Completion = struct {
+        io: *IO,
+        result: i32 = undefined,
+        next: ?*Completion = undefined,
+        operation: Operation,
+        // This is one of the usecases for c_void outside of C code and as such c_void will
+        // be replaced with anyopaque eventually: https://github.com/ziglang/zig/issues/323
+        context: ?*c_void,
+        callback: fn (context: ?*c_void, completion: *Completion, result: *const c_void) callconv(.C) void,
+
+        fn prep(completion: *Completion) void {
+            completion.io.get_sqe(completion);
+        }
+
+        fn complete_prep(completion: *Completion, sqe: *io_uring_sqe) void {
```

Comment:
Maybe `prep()` should be `queue()` or `enqueue()` and `complete_prep` should be `prep` because that's what they're doing?

Reply:
Yeah that makes sense, as you may have noticed I struggled a bit with naming things here :)

Reply:
No worries, I actually think you improved most of the names, FIFO, in/out/push etc.